### DISCUSSION TITLE
cmake is not triggered on build command

### DISCUSSION THREAD

**Discussion by Nordsoft91** [axmolengine/axmol#1551] [Help Desk]
- axmol version: axmol-2.1.0-rc7
- devices test on: MacOS M1
- developing environments
   - Xcode version: 14.2
   - cmake version: 3.27.2
Steps to Reproduce:
1. Create new project
1. Execute `axmol build -p osx -a x64 -configOnly`
2. Go to `build_x64` folder and open .xcodeproj with XCode
3. Create some new .cpp file in `Source` folder
4. Execute `axmol build -p osx -a x64 -configOnly` again

Expected: xcode project is updated, new source file is added
Actual: nothing changed, to get actual update you must delete CMakeCache.txt and do full rebuild

**Comment by rh101** [comment]
If you're using `GLOB` or `GLOB_RECURSE` in your project `CMakeLists.txt`, then that would most likely be the source of the problem, and has nothing to do with Axmol.

Check this section in the file:
```
file(GLOB_RECURSE GAME_HEADER
    Source/*.h
    )
file(GLOB_RECURSE GAME_SOURCE
    Source/*.cpp
    )
```

From the CMake [docs](https://cmake.org/cmake/help/latest/command/file.html#filesystem):
![image](https://github.com/axmolengine/axmol/assets/8603230/e5c947ad-cba9-4382-9249-92aa013ccaa2)

It only takes you a few seconds to manually add a new cpp/h file reference to `CMakeLists.txt`, so save yourself the hassle and avoid using `GLOB`/`GLOB_RECURSE`.

**Comment by Nordsoft91** [comment]
Yes, but CMakeLists.txt is generated by axmol automatically and it's quite big. Possible solution would be to generate short CMakeLists.txt in `Source` directory automatically which will be much shorter and contain (almost) those lines and include it in root CMakeLists.txt. Thus will be much easier to work with it.

**Comment by Nordsoft91** [comment]
I replaced adding of sources in root CMakeLists.txt:
```
add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/Source)
```

Now `Source/CMakeLists.txt`:
```
# The common cross-platforms source files and header files
file(GLOB_RECURSE GAME_HEADER
    *.h
    )
file(GLOB_RECURSE GAME_SOURCE
    *.cpp
    )

set(GAME_HEADER ${GAME_HEADER} PARENT_SCOPE)
set(GAME_SOURCE ${GAME_SOURCE} PARENT_SCOPE)
```

**Comment by halx99** [comment]
Try add param: `-force`

**Comment by rh101** [comment]
> Yes, but CMakeLists.txt is generated by axmol automatically and it's quite big. Possible solution would be to generate short CMakeLists.txt in `Source` directory automatically which will be much shorter and contain (almost) those lines and include it in root CMakeLists.txt. Thus will be much easier to work with it.

That is something developers would most likely be doing already, but it may help to add it to the project templates just to make things a little clearer. Consider submitting a PR with the change.
